code/parachain/frame/fnft/src/lib.rs

```diff
@@ -471,14 +475,17 @@ pub mod pallet {
 	}
 
 	impl<T: Config> FinancialNft<AccountIdOf<T>> for Pallet<T> {
-		/// TODO (vim): Assess the probability of collision in generating accounts
-		///   with collection id type u128 and instance id type u128 this definitely collides
-		///   because the seed is longer than the account ID
 		fn asset_account(
 			collection: &Self::CollectionId,
 			instance: &Self::ItemId,
 		) -> AccountIdOf<T> {
-			T::PalletId::get().into_sub_account_truncating((collection, instance))
+			// `into_sub_account_truncating()` gives us 20 bytes of space to create a seed.
+			// `blake2_256()` returns 32 bytes of data, however, BLAKE2 already truncates its
+			// results. Truncating this to 20 bytes puts us at the recomended output length for
+			// BLAKE2.
+			T::PalletId::get().into_sub_account_truncating(sp_io::hashing::blake2_256(
```

Comment:
not sure why you hash if you already have hash for `(collection, instance)` ? weird

Reply:
Does [hashed_key_for](https://paritytech.github.io/substrate/master/frame_support/pallet_prelude/struct.StorageDoubleMap.html#method.hashed_key_for) not recompute the hash each time anyway? @dzmitry-lahoda

Reply:
Also, that key would be prefixed, which would be even more bytes to be concerned about creating collisions for. We'd *probably* still be safe, but not with the same probability

Reply:
it stores key. not rehashes each time.

Reply:
not sure, 
in both cases hash in put `collection, instance`. given same amount of data given hashes are good enough you will not doo better. 

PalletId ads it name as prefix to account to any hash you create. so in any case hash_out_of_curve(palletid, collection, instance) = account_id.

So in case of existing case, there is no compute and easy migration rules.

Reply:
Yes, this also recomputes the hash and includes extra data that we would have to trim.
See underlying function [here](https://github.com/paritytech/substrate/blob/a3aa0e8f7bc8b10b981eb634f9e257f075e9aa15/frame/support/src/storage/generator/map.rs#L66~L80)

Reply:
holly...

Reply:
so when you when you check if item exists you get hash. and can reuse? so really you still better as only 1 hash i stead of 2. so hash can be captured from earliercomputstion. a d weird that by spec it stores hash and binary data glued. issue is that there is no reverse map.

Reply:
do you check item exists before call to asset_account? which would calculate hash? so really can just cal ulate hash and check it exists and then do.subaccount.

also, it adds storage prefix which may be improves hash collision?

Reply:
> do you check item exists before call to asset_account? which would calculate hash? so really can just cal ulate hash and check it exists and then do.subaccount.
No, I'm not assuming that the fNFT exist before this function is called. I only assume the `collection` and `instance` are potentially valid.
> PalletId ads it name as prefix to account to any hash you create. so in any case hash_out_of_curve(palletid, collection, instance) = account_id.
Not necessarily. 
Currently `hash_out_of_curve(palletIdType, palletId, hash(collection, instance)) => accountId`.
What I think you're recommending would result in `hash_out_of_curve(palletTypeId, palletId, storageKey) => accountId`. Where `storageKey` is `palletId(8-byte) + blake128(collection)(16byte) + collection(16-byte) + blake128(instance)(16-byte) + instance(8-byte)`.
So, we would only capture `palletId` and part of `blake128(collection)` within the 20 bytes we have. I think that would guarantee collisions.

Reply:
[you no need to call into_account. you can just do Account32::from](https://github.com/paritytech/substrate/blob/20b5aac7e09e8b25abaa48c6781e653b36ba3ca0/client/transaction-pool/src/graph/pool.rs#L467)

Reply:
or other form of from. so what Pallet::into_subaccount does?

Reply:
just do not do it

Reply:
so if you have 32 bit and hash into 64 bit. and then hash into 32 back it still will be good if hash is good. so it should not loose quality if it expands and then contracts back.

Reply:
> storageKey is palletId(8-byte) + blake128(collection)(16byte) + collection(16-byte) + blake128(instance)(16-byte) + instance(8-byte)

that is not true

Reply:
```
	fn storage_map_final_key<KeyArg>(key: KeyArg) -> Vec<u8>
	where
		KeyArg: EncodeLike<K>,
	{
		let storage_prefix = storage_prefix(Self::module_prefix(), Self::storage_prefix());
		let key_hashed = key.borrow().using_encoded(Self::Hasher::hash);

		let mut final_key = Vec::with_capacity(storage_prefix.len() + key_hashed.as_ref().len());

		final_key.extend_from_slice(&storage_prefix);
		final_key.extend_from_slice(key_hashed.as_ref());

		final_key
	}
```

Reply:
Account::from(final_key) where final_key is storage_prefix + key_hashed. where storage_prefix = hash(pallet + storage) and key_hashed = hash(collection + item id).

account::from(hash(hash(pallet, storage), hash(collection, id))) ?

Reply:
@PoisonPhang I bet 10 PICA Account32 is 32 byte and storage_map_final_key returns 32 byte.  So you do not even need to hash and have account id = key => can check existence easy when needed without storing reverse map on top (as will happen with suggested approach)

Reply:
and you as conflict free as substrate itself conflic free as it uses it for KV in rocksdb - I still suggest you to read Tabrizi blog and watch video))))

Reply:
So, you're suggesting we skip the [into_sub_account_truncating](https://paritytech.github.io/substrate/master/sp_runtime/traits/trait.AccountIdConversion.html#tymethod.into_sub_account_truncating) and derive an account from the full storage key? I misunderstood
This feels byzantine AF, but it could work.

> that is not true

I think it is true. I sent the function for a StorageMap before, but we use a StorageDoubleMap which uses [this function](https://github.com/paritytech/substrate/blob/a3aa0e8f7bc8b10b981eb634f9e257f075e9aa15/frame/support/src/storage/generator/double_map.rs#L92~L111) and currently use Blake2 128 Concat for the keys. 

Concat outputs blake2_128(value) + value

Reply:
> and you as conflict free as substrate itself conflic free as it uses it for KV in rocksdb - I still suggest you to read Tabrizi blog and watch video))))

I got this lecture live from Shawn himself. What you're proposing should work, it just feels byzantine. We can try it tho.

Reply:
> I think it is true.

In the context of what I assumed before, using `into_sub_`

Reply:
> [this function](https://github.com/paritytech/substrate/blob/a3aa0e8f7bc8b10b981eb634f9e257f075e9aa15/frame/support/src/storage/generator/double_map.rs#L92~L111),

id does `hash(hash(pallet, storage), hash(collection, id)`

parity says that if you control ids, via incrementing, it is "good" to use weaker hashes like 64 bit. 

so instead of 128 + 128 + 128 you can use do 128(I assume pallet + storage hash 128 which may be less) + 64 + 64. 
if prefix is 64 , can do 64 + 128 + 64.  

`into_sub_` hashes hash. so it does not gets new information into it. not sure why it should be considered more collision resistant if there is no new if collision resistance = length of output + information in input(you do not seed changes which fight againts hash for collision) + hash function quality i guess. not expert :) . i guess all are good at case. hash(hash(x)) may not be needed.

Reply:
Sir, I'll take my 10 PICA @dzmitry-lahoda 
```rust
fn asset_account(
			collection: &Self::CollectionId,
			instance: &Self::ItemId,
		) -> AccountIdOf<T> {
			println!("{:?}", Instance::<T>::hashed_key_for(collection, instance).len());
			// `into_sub_account_truncating()` gives us 20 bytes of space to create a seed.
			// `blake2_256()` returns 32 bytes of data, however, BLAKE2 already truncates its
			// results. Truncating this to 20 bytes puts us at the recomended output length for
			// BLAKE2.
			T::PalletId::get()
				.into_sub_account_truncating(sp_io::hashing::blake2_256(&(1).encode()))
		}
```
```
88
88
thread 'test::nonfungibles::transfer::roundtrip' panicked at 'called `Result::unwrap()` on an `Err` value: Module(ModuleError { index: 4, error: [4, 0, 0, 0], message: Some("Duplicate") })', parachain/frame/fnft/src/test/prelude.rs:98:74
```

Reply:
This may change if we use a weaker hash function as you stated, blake2_128_concat is overkill in this situation.
> parity says that if you control ids, via incrementing, it is "good" to use weaker hashes like 64 bit.